pkg/operation/botanist/secrets.go

```diff
@@ -526,21 +526,38 @@ func (b *Botanist) DeleteOldServiceAccountSecrets(ctx context.Context) error {
 		log := b.Logger.WithValues("serviceAccount", client.ObjectKeyFromObject(&serviceAccount))
 
 		taskFns = append(taskFns, func(ctx context.Context) error {
-			// If the ServiceAccount has none or only one secret then there is nothing left to clean up. Otherwise, we
-			// should drop all secrets except for the first one in the list (which is the most recent secret signed with
-			// the new token key).
-			if len(serviceAccount.Secrets) <= 1 {
-				return nil
+			// Wait until we are allowed by the limiter to not overload the kube-apiserver with too many requests.
+			if err := limiter.Wait(ctx); err != nil {
+				return err
 			}
 
-			var secretsToDelete []client.Object
-			for _, secretReference := range serviceAccount.Secrets[1:] {
-				secretsToDelete = append(secretsToDelete, &corev1.Secret{ObjectMeta: metav1.ObjectMeta{Name: secretReference.Name, Namespace: serviceAccount.Namespace}})
+			var (
+				secretsToDelete  []client.Object
+				remainingSecrets []corev1.ObjectReference
+			)
+
+			// In the CreateNewServiceAccountSecrets function we add a new ServiceAccount secret as the first one to the
+			// .secrets[] list in the ServiceAccount resource. However, when we reach this code now, the user could have
+			// already removed this secret or changed the .secrets[] list. Hence, we now check which of the secrets in
+			// the list have been created before the credentials rotation completion has been triggered. We only delete
+			// those and keep the rest of the list untouched to not interfere with the user's operations.
+			for _, secretReference := range serviceAccount.Secrets {
+				secret := &corev1.Secret{}
+				if err := b.ShootClientSet.Client().Get(ctx, client.ObjectKey{Name: secretReference.Name, Namespace: serviceAccount.Namespace}, secret); err != nil {
```

Comment:
This could be a metadata-only get.

Reply:
https://github.com/gardener/gardener/pull/7336